routing/cross_mwm_graph.cpp

```diff
@@ -0,0 +1,201 @@
+  , m_crossMwmIndexGraph(index, numMwmIds)
+  , m_crossMwmOsrmGraph(numMwmIds, indexManager)
+{
+  CHECK(m_numMwmIds, ());
+  CHECK(m_vehicleModelFactory, ());
+}
+
+bool CrossMwmGraph::IsTransition(Segment const & s, bool isOutgoing)
+{
+  // Index graph based cross-mwm information.
+  if (CrossMwmSectionExists(s.GetMwmId()))
+    return m_crossMwmIndexGraph.IsTransition(s, isOutgoing);
+  return m_crossMwmOsrmGraph.IsTransition(s, isOutgoing);
+}
+
+void CrossMwmGraph::GetTwins(Segment const & s, bool isOutgoing, vector<Segment> & twins)
+{
+  CHECK(IsTransition(s, isOutgoing),
+        ("The segment", s, "is not a transition segment for isOutgoing ==", isOutgoing));
+  // Note. There's an extremely rare case when a segment is ingoing and outgoing at the same time.
+  // |twins| is not filled for such cases. For details please see a note in
+  // CrossMwmGraph::GetEdgeList().
+  if (IsTransition(s, !isOutgoing))
+    return;
+
+  twins.clear();
+
+  // Note. The code below looks for twins based on geometry index. This code works for
+  // any combination mwms with different cross mwm sections.
+  // It's possible to implement a faster version for two special cases:
+  // * all neighboring mwms have cross_mwm section
+  // * all neighboring mwms have osrm cross mwm sections
+  TransitionPoints const points = GetTransitionPoints(s, isOutgoing);
+  for (m2::PointD const & p : points)
+  {
+    // Node. The map below is necessary because twin segments could belong to several mwm.
+    // It happens when a segment crosses more than one feature.
+    map<NumMwmId, ClosestSegment> minDistSegs;
+    auto const findBestTwins = [&](FeatureType & ft) {
+      if (ft.GetID().m_mwmId.GetInfo()->GetType() != MwmInfo::COUNTRY)
+        return;
+
+      if (!ft.GetID().IsValid())
+        return;
+
+      NumMwmId const numMwmId =
+          m_numMwmIds->GetId(ft.GetID().m_mwmId.GetInfo()->GetLocalFile().GetCountryFile());
+      if (numMwmId == s.GetMwmId())
+        return;
+
+      if (!m_vehicleModelFactory->GetVehicleModelForCountry(ft.GetID().GetMwmName())->IsRoad(ft))
+        return;
+
+      ft.ParseGeometry(FeatureType::BEST_GEOMETRY);
+      vector<Segment> twinCandidates;
+      GetTwinCandidates(ft, !isOutgoing, twinCandidates);
+      for (Segment const & tc : twinCandidates)
+      {
+        TransitionPoints const twinPoints = GetTransitionPoints(tc, !isOutgoing);
+        for (m2::PointD const & tp : twinPoints)
+        {
+          double const distM = MercatorBounds::DistanceOnEarth(p, tp);
+          if (distM == 0.0)
+          {
+            twins.push_back(tc);
+            minDistSegs[numMwmId].m_exactMatchFound = true;
+          }
+          if (!minDistSegs[numMwmId].m_exactMatchFound && distM <= kTransitionEqualityDistM &&
+              distM < minDistSegs[numMwmId].m_bestDistM)
+          {
+            minDistSegs[numMwmId].m_bestDistM = distM;
+            minDistSegs[numMwmId].m_bestSeg = tc;
```

Comment:
Done. And it refactored as well. Now there's a method Update at class ClosestSegment:
```c++
    TransitionPoints const twinPoints = GetTransitionPoints(tc, !isOutgoing);
    for (m2::PointD const & tp : twinPoints)
    {
      double const distM = MercatorBounds::DistanceOnEarth(point, tp);
      ClosestSegment & closestSegment = minDistSegs[numMwmId];
      double constexpr kEpsMeters = 2.0;
      if (my::AlmostEqualAbs(distM, 0.0, kEpsMeters))
      {
        twins.push_back(tc);
        closestSegment.m_exactMatchFound = true;
      }
      closestSegment.Update(distM, tc);
    }
```